jax/_src/clusters/mpi4py_cluster.py

```diff
@@ -0,0 +1,130 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from jax._src import clusters
+import socket
+
+from numpy import unique, where
+
+from importlib.util import find_spec
+
+class Mpi4pyCluster(clusters.ClusterEnv):
+  
+
+  name: str = "mpi4py"
+  opt_in_only_method: bool = True
+
+  @classmethod
+  def is_env_present(cls) -> bool:
+    # Why include and opt_in?  Enables this class to conform to
+    # every other ClusterEnv subclass while always being rejected
+    # as viable, except in the express case where we request to check
+    # it explicitly.
+
+    # in many HPC clusters, the variables `https_proxy` and `http_proxy`
+    # are set to enable access to normally unreachable network locations.
+    # For example, `pip install ...` fails on compute nodes without them.
+
+    # Unfortunately, these variables break the jax distributed init.
+    # The user needs to unset them, but I don't want to modify the global
+    # python os.environ here for them (bad practice)
+
+    # And I also don't know what the right way to raise a complaint here is
+    
+    # Relies on mpi4py:
+    return find_spec("mpi4py") is not None
+
+  @classmethod
+  def get_coordinator_address(cls, timeout_secs: int | None) -> str:
+
+    # Using mpi4py, figure out rank 0 and it's hostname.
+    # Then broadcast the hostname and port.
+
+
+    from mpi4py import MPI
+    # Get the global communicator:
+    COMM_WORLD = MPI.COMM_WORLD
+
+    # On rank 0, get the hostname:
+
+    if COMM_WORLD.Get_rank() == 0:
+        # Order all the hostnames, and find unique ones
+        hostname = socket.gethostname()
+
+        # Apparently, we want to pick a port in an ephemeral range...
+        port_id = hash(hostname) % 2**12 + (65535 - 2**12 + 1)
+        
+        hostname = f'{hostname}:{port_id}'
+
+    else:
+        hostname = None
+   
+
+
+    # Broadcast the host_ip to all ranks:
+    hostname = COMM_WORLD.bcast(hostname, root=0)
+    # host_ip = COMM_WORLD.bcast(host_ip, root=0)
+
+
+    return hostname
+
+
+  @classmethod
+  def get_process_count(cls) -> int:
+    from mpi4py import MPI
+    return int(MPI.COMM_WORLD.Get_size())
+
+  @classmethod
+  def get_process_id(cls) -> int:
+    from mpi4py import MPI
+    return int(MPI.COMM_WORLD.Get_rank())
+  
+  @classmethod
+  def get_local_process_id(cls) -> int | None:
+
+    # Using mpi4py, split the global communicator into sub communicators
+    # based on hostname.  mpi will assign them ranks and that will allow
+    # a selection of the local process ID.
+    from mpi4py import MPI
+    COMM_WORLD = MPI.COMM_WORLD
+
+    # This is a previous method that is replaced with a different mpi split:
+
+    # hostname = socket.gethostname()
+    # # host_key = host_key %
+    # all_hostnames = COMM_WORLD.gather(hostname, root=0)
+
```

Comment:
nit: Could you remove the commented out code since it is replace by the split method?